src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,219 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import data.polynomial.monomial
+import data.matrix.pequiv
+import data.equiv.fin
+
+/-!
+# Circulant matrices
+
+This file contains the definition and basic results about circulant matrices.
+Given a vector `v : I → α` indexed by a type that is endowed with subtraction,
+`circulant v` is the matrix whose `(i,j)`th entry is `v (i = j)`.
+
+## Main results
+
+- `matrix.circulant`: the circulant matrix generated by a given vector `v : I → α`.
+
+## Implementation notes
+
+`matrix.fin.foo` is the `fin n` version of `matrix.foo`.
+Namely, the index type of the circulant matrices in discussion is `fin n`.
+
+## Tags
+
+circulant, matrix
+-/
+
+variables {α β I J R : Type*} {n : ℕ}
```

Comment:
Can you change `I J` to `m n` to match the convention we use elsewhere? I'd suggest you then change `n : ℕ` to `k : ℕ` to avoid the clash

Reply:
Won't using lowercase and adjacent letters to denote both index types and a natural number be confusing? @eric-wieser 
I used the letters now in the file to avoid possible confusion.

Reply:
Perhaps use `n' : ℕ` then instead? I think it's better to be consistent with the other matrix files by having the default indices be called `m` and `n`.

Reply:
If using `n'`. One may still think `n` and `n'` denote the same type of things. Should such possibility of causing confusion be OK?

Reply:
I personally prefer to use `I` `J` in this file to avoid any possible confusion. The intention for `I` `J` `n` seems clearer.

Reply:
Frankly I think you should just remove `n` here; I think almost none of your lemmas below use it:
```suggestion
variables {α β m n R : Type*}
```

Reply:
> I personally think the intention for using `I` `J` `n` in this file seems clearer.

If every author does what _they_ think is clearer, you end up with lots of files that don't match, which makes it annoying when you have to move lemmas between files. It's usually way more important to be consistent with what old authors already decided than it is to do what would be best in isolation.

If you feel _really_ strongly about this, you could make a follow-up PR to rename `m n` to `I J` in all the matrix files; but I expect you would be told that `matrix I J R` is worse because a mathematician expects to talk about `m x n` matrices.